nexus/src/app/background/plan_execution.rs

```diff
@@ -0,0 +1,408 @@
+            creator: "test".to_string(),
+            comment: "test blueprint".to_string(),
+        }
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_deploy_omicron_zones(cptestctx: &ControlPlaneTestContext) {
+        let nexus = &cptestctx.server.apictx().nexus;
+        let datastore = nexus.datastore();
+        let opctx = OpContext::for_tests(
+            cptestctx.logctx.log.clone(),
+            datastore.clone(),
+        );
+
+        let (blueprint_tx, blueprint_rx) = watch::channel(None);
+        let mut task = PlanExecutor::new(datastore.clone(), blueprint_rx);
+
+        // With no blueprint we should fail with an appropriate message.
+        let value = task.activate(&opctx).await;
+        assert_eq!(value, json!({"error": "no blueprint"}));
+
+        // Get a success (empty) result back when the blueprint has an empty set of zones
+        let blueprint = Arc::new(create_blueprint(BTreeMap::new()));
+        blueprint_tx.send(Some(blueprint)).unwrap();
+        let value = task.activate(&opctx).await;
+        assert_eq!(value, json!({}));
+
+        // Create some fake sled-agent servers to respond to zone puts and add
+        // sleds to CRDB.
+        let mut s1 = httptest::Server::run();
+        let mut s2 = httptest::Server::run();
+        let sled_id1 = Uuid::new_v4();
+        let sled_id2 = Uuid::new_v4();
+        let rack_id = Uuid::new_v4();
+        for (i, (sled_id, server)) in
+            [(sled_id1, &s1), (sled_id2, &s2)].iter().enumerate()
+        {
+            let SocketAddr::V6(addr) = server.addr() else {
+                panic!("Expected Ipv6 address. Got {}", server.addr());
+            };
+            let update = SledUpdate::new(
+                *sled_id,
+                addr,
+                SledBaseboard {
+                    serial_number: i.to_string(),
+                    part_number: "test".into(),
+                    revision: 1,
+                },
+                SledSystemHardware {
+                    is_scrimlet: false,
+                    usable_hardware_threads: 4,
+                    usable_physical_ram: ByteCount(1000.into()),
+                    reservoir_size: ByteCount(999.into()),
+                },
+                rack_id,
+            );
+            datastore
+                .sled_upsert(update)
+                .await
+                .expect("Failed to insert sled to db");
+        }
+
+        // The particular dataset doesn't matter for this test.
+        // We re-use the same one to not obfuscate things
+        let dataset = OmicronZoneDataset {
+            pool_name: format!("oxp_{}", Uuid::new_v4()).parse().unwrap(),
+        };
+
+        let generation = Generation::new();
+
+        // Zones are updated in a particular order, but each request contains
+        // the full set of zones that must be running.
+        // See https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/rack_setup/service.rs#L976-L998
```

Comment:
Linking to `main` will atrophy over time, but I'm not sure I like a link to a specific commit much better. Could we reference a method name instead, maybe? If that changes there's a chance this comment gets updated when someone is searching for the method name.

Reply:
Fixed in 24b34f70f76a5795d51987442f392dc374cc71eb